lib/include/aws_iot_defender.h

```diff
@@ -21,45 +21,84 @@
 
 /**
  * @file aws_iot_defender.h
- * @brief
+ * @brief User-facing functions and structs of AWS IoT Device Defender libraries
+ *
+ * [Link to AWS documentation](https://docs.aws.amazon.com/iot/latest/developerguide/device-defender-detect.html)
+ *
  */
 
 #ifndef _AWS_IOT_DEFENDER_H_
 #define _AWS_IOT_DEFENDER_H_
 
+/* Build using a config header, if provided. */
+#ifdef AWS_IOT_CONFIG_FILE
+    #include AWS_IOT_CONFIG_FILE
+#endif
+
 /* Standard includes. */
 #include <stdint.h>
 
 /* MQTT include. */
 #include "aws_iot_mqtt.h"
 
-/* Network include. */
+/* Platform network include. */
 #include "platform/aws_iot_network.h"
 
+#define AWS_IOT_DEFENDER_METRICS_ALL                          0xffffffff /**< @brief Flag to indicate including all metrics. */
+
+#define AWS_IOT_DEFENDER_METRICS_TCP_CONN_ESTA_TOTAL          ( 1 << 0 ) /**< @brief Total count of established TCP connections. */
+#define AWS_IOT_DEFENDER_METRICS_TCP_CONN_ESTA_LOCAL_PORT     ( 1 << 1 ) /**< @brief Local port number of established TCP connections. */
+#define AWS_IOT_DEFENDER_METRICS_TCP_CONN_ESTA_REMOTE_ADDR    ( 1 << 2 ) /**< @brief Remote address (IP:port) of established TCP connections. For example, 192.168.0.1:8000. */
+
+/**
+ * @brief Connections metrics including local port number and remote address.
+ */
+#define AWS_IOT_DEFENDER_METRICS_TCP_CONN_ESTA_CONNS                                                           \
+    ( AWS_IOT_DEFENDER_METRICS_TCP_CONN_ESTA_LOCAL_PORT | AWS_IOT_DEFENDER_METRICS_TCP_CONN_ESTA_REMOTE_ADDR ) \
+
+
+/**
+ * @brief Established connections metrics including connections metrics and total count.
+ */
+#define AWS_IOT_DEFENDER_METRICS_TCP_CONN_ESTA                                                      \
+    ( AWS_IOT_DEFENDER_METRICS_TCP_CONN_ESTA_CONNS | AWS_IOT_DEFENDER_METRICS_TCP_CONN_ESTA_TOTAL ) \
+
+#define AWS_IOT_DEFENDER_METRICS_LIS_TCP_TOTAL         ( 1 << 0 ) /**< @brief Total count of listening TCP ports. */
+#define AWS_IOT_DEFENDER_METRICS_LIS_TCP_PORTS_PORT    ( 1 << 1 ) /**< @brief Port number of listening TCP ports. */
+
+/**
+ * @brief Ports metrics including port number.
+ */
+#define AWS_IOT_DEFENDER_METRICS_LIS_TCP_PORTS         ( AWS_IOT_DEFENDER_METRICS_LIS_TCP_PORTS_PORT )
+
+
+#define AWS_IOT_DEFENDER_START_INFO_INITIALIZER        { 0 }
+
 /**
- * This enum type is the ID of metrics. It can be specified in AwsIotDefender_SetMetrics function.
+ * @brief Metrics group options for AwsIotDefender_SetMetrics() function.
  */
 typedef enum
 {
-    AWS_IOT_DEFENDER_METRIC_TCP_ESTABLISHED_TOTAL,
-} AwsIotDefenderMetric_t;
+    AWS_IOT_DEFENDER_METRICS_TCP_CONN,
+    AWS_IOT_DEFENDER_METRICS_LIS_TCP,
```

Comment:
What does LIS stands for? Not very intuitive...

Reply:
LISTENING? Better long names...  aldo make CONN into CONNECTION

Reply:
I like long time. I was just concerned if the macro name too long will cause problem..

Reply:
I like long *name*